### PR TITLE
 Support PathFilter when fetching splits for Hudi tables

### PR DESCRIPTION
**Pull request by bhasudha**
This PR is corresponds to the issue - https://github.com/prestodb/presto/issues/13511
Summary:
    - Introduce PathFilter support in DirectoryLister interface and HiveFileIterator
    - Unit test HiveFileIterator using PathFilter
    - Plug specific PathFilter implementation for HoodieParquetInputFormat

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==

General Changes
* ...
* ...

Hive Changes
* ...
* ...
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by bhasudha** [comment]
@arhimondr  please take a look when you can. Thanks much!

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 1]
```diff
@@ -154,11 +158,13 @@ public BackgroundHiveSplitLoader(
         this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
         this.namenodeStats = requireNonNull(namenodeStats, "namenodeStats is null");
         this.directoryLister = requireNonNull(directoryLister, "directoryLister is null");
+        this.pathFilter = pathFilterProvider.getPathFilter();
         this.recursiveDirWalkerEnabled = recursiveDirWalkerEnabled;
         this.executor = requireNonNull(executor, "executor is null");
         this.partitions = new ConcurrentLazyQueue<>(requireNonNull(partitions, "partitions is null"));
-        this.hdfsContext = new HdfsContext(session, table.getDatabaseName(), table.getTableName());
+        this.hdfsContext = new HdfsEnvironment.HdfsContext(session, table.getDatabaseName(), table.getTableName());
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: nit: let's revert this change

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import org.apache.hadoop.fs.PathFilter;
+
+public class DefaultPathFilterProvider
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: Or maybe `HadoopPathFilterProvider` since hadoop is the default storage? I don't have strong opinion here.

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        implements PathFilterProvider
+{
+    private final PathFilter pathFilter;
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: Let's make this a static field and directly return when `getPathFilter` is called.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveBenchmarkQueryRunner.java [hunk 4]
```diff
@@ -77,7 +77,7 @@ public static LocalQueryRunner createLocalQueryRunner(File tempDir)
         HiveConnectorFactory hiveConnectorFactory = new HiveConnectorFactory(
                 "hive",
                 HiveBenchmarkQueryRunner.class.getClassLoader(),
-                Optional.of(metastore));
+                Optional.of(metastore), new DefaultPathFilterProviderModule());
```
Comment (shixuan-fan) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveBenchmarkQueryRunner.java]: nit: line break between arguments.

Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/HiveBenchmarkQueryRunner.java]: done!

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java (hunk 3 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: sure

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 1 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: done!

**Comment by wenleix** [comment]
@bhasudha : Assume this is ready for review, I will remove the `[WIP]` in the title.

**Comment by bhasudha** [comment]
> @bhasudha : Assume this is ready for review, I will remove the `[WIP]` in the title.

@wenleix I kept it as WIP since I am still testing it. I have some related questions. When I bring up local presto server, I believe presto-hudi tries to register two connectors - hudi and hive (since its extending the HiveHadoop2Plugin). I see error saying hive connector is already registered. I am wondering if this is got to do with the way I implemented the presto-hudi connector.  Should I change HudiPlugin to extend from HivePlugin like how HiveHadoop2Plugin does instead of extending HiveHadoop2Plugin? Do you have any recommendations around this?

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import org.apache.hadoop.fs.PathFilter;
+
+public class DefaultPathFilterProvider
+        implements PathFilterProvider
+{
+    private static PathFilter pathFilter = path -> true;
```
Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: nit: final

**Comment by shixuan-fan** [comment]
@bhasudha In our internal extension of hive connector, we didn't extend `HivePlugin`. We just implemented `Plugin`, and return a `HiveConnector` in `ConnectorFactory#create`.

File: presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java [hunk 6]
```diff
@@ -13,11 +13,21 @@
...
+
 public class HiveHadoop2Plugin
         extends HivePlugin
 {
     public HiveHadoop2Plugin()
     {
         super("hive-hadoop2");
     }
+
+    public HiveHadoop2Plugin(String name, Optional<ExtendedHiveMetastore> metastore, Module pathProviderModule)
```
Comment (arhimondr) [inline presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java]: `Optional<ExtendedHiveMetastore> metastore` is effectively unused.

Comment (arhimondr) [inline presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java]: Why does the `PathFilterProvider` has to be wrapped in a module? Why not simply supply the `PathFilterProvider` itself?

e.g.: change the interface to `public HiveHadoop2Plugin(String name, Optional<PathFilterProvider> pathFilterProvider)`

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java (hunk 5 above)
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: I would simply inline it.

File: presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import org.apache.hadoop.fs.PathFilter;
+
+public interface PathFilterProvider
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java]: Let's rename it to `FileFormatSpecificPathFilterFactory`, as it always creates a new one.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 8]
```diff
@@ -359,9 +366,13 @@ private void invokeNoMoreSplitsIfNecessary()
...
-        // To support custom input formats, we want to call getSplits()
-        // on the input format to obtain file splits.
-        if (shouldUseFileSplitsFromInputFormat(inputFormat)) {
+        // Specifically for custom InputFormats:
+        // We want to call getSplits() on the input format to obtain file splits. This requires the InputFormat
+        // implementation to be annotated with UseFileSplitsFromInputFormat. While this is a preferred behavior,
+        // custom input formats can also pass in a PathFilter implementation that will be used with regular Presto's
+        // Directory Listing operation. To opt this route, disable 'hive.use-file-splits-from-input-format' and bind
+        // PathFilterProvider implementation to PathFilterProvider interface instead of DefaultPathFilterProvider.
+        if (shouldUseFileSplitsFromInputFormat(useFileSplitsFromInputPath, inputFormat)) {
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Having a separate property looks weird.

Let's change the organization a little bit more.

First of all - the `PathFilterProvider` can be changed

```
public interface PathFilterProvider
{
    PathFilter getPathFilter();
}
```

->

```
public interface FileFormatSpecificPathFilterFactory
{
    Optional<PathFilter> createPathFilter(String inputFormatName);
}
```

Then, if the `FileFormatSpecificPathFilterFactory` returns a `PathFilter` for a given format, we assume it is knowingly overrides the default behaviour (e.g.: somebody, who attached that knows what they are doing), and the `shouldUseFileSplitsFromInputFormat` is ignored for that format.

File: presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java (hunk 7 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java]: Also please leave a comment on that interface explaining what it is for. 

It might not be trivial for somebody who is not in the loop to understand that.

File: presto-hudi/pom.xml [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            <groupId>com.facebook.presto</groupId>
+            <artifactId>presto-hive-hadoop2</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>com.facebook.presto</groupId>
+            <artifactId>presto-hive</artifactId>
+        </dependency>
+
+        <dependency>
```
Comment (arhimondr) [inline presto-hudi/pom.xml]: When I suggested to extract a separate module - I thought that Hudi is something that is not open source.

Since it is opensource, is there's a good reason why don't we want to just have it included directly in `presto-hive`  and entirely avoid unnecessary interfacing?

CC: @wenleix @shixuan-fan

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java (hunk 5 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: I removed this class in the most recent patch.

File: presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java (hunk 6 above)
Comment (bhasudha) [reply presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java]: I removed changes to this plugin.

Comment (bhasudha) [reply presto-hive-hadoop2/src/main/java/com/facebook/presto/hive/HiveHadoop2Plugin.java]: Done!

File: presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java (hunk 5 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/DefaultPathFilterProvider.java]: sure.

File: presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java (hunk 7 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/PathFilterProvider.java]: Done!

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 8 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: sounds good.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 10]
```diff
@@ -130,6 +133,7 @@
     private final ReentrantReadWriteLock taskExecutionLock = new ReentrantReadWriteLock();
 
     private HiveSplitSource hiveSplitSource;
+    private AtomicReference<Optional<PathFilter>> pathFilterReference = new AtomicReference<>();
```
Comment (bhasudha) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: I need this AtomicReference since loadingPartitions is a multi threaded implementation. Ideally  the first thread that processes the partitions would set the PathFilter and rest of the threads can simply use it. I kept it as AtomicReference since we wanted the FileFormatSpecificPathFilterFactory.createPathFilter(inputFormatName) to take in a String inputFormatName as a parameter. To identify this inputformat name inside the  BackgroundSplitLoader constructor may be a bit hacky. Instead if the createPathFilter(inputFormatName) can be changed to getPathFilter() then we wouldnt need the AtomicReference (We can directly assign the PathFilter in the BackgroundSplitLoader constructor). Please let me know if you prefer one way or other.

File: presto-hudi/pom.xml (hunk 9 above)
Comment (bhasudha) [reply presto-hudi/pom.xml]: Hudi is opensourced and is indeed supported by Google Cloud and AWS EMR :) 

Regarding your comment on moving this back to presto-hive, we could still continue with presto-hudi as a connector module making it a placeholder for longterm. Any further feature requests from Hudi on Presto side can levergae this.  But please let me know your thoughts.

Comment (arhimondr) [reply presto-hudi/pom.xml]: From what I understand, at the end of the day Hudi is just yet another table format. It is quite similar to ORC / Parquet, but apparently a little bit broader scoped as it also specifies the way files have to be organized.

Since Parquet / ORC are first class citizens in `presto-hive` i wonder why Hudi cannot be so?
Eventually we may want to introduce Hudi as a first class citizen allowing table writes in this format. We may want to add it as an option to our list of officially supported formats specified by the https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/HiveStorageFormat.java#L42

@wenleix @shixuan-fan @highker Do you have any objections against including Hudi into presto-hive module?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 10 above)
Comment (arhimondr) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: The `BackgroundHiveSplitLoader` loads splits for multiple partitions. Partitions are allowed to have distinct storage format. Why not simply recreated the filter for every partition? Is it not a lightweight operation enough?

File: presto-hudi/pom.xml (hunk 9 above)
Comment (haoyuan) [reply presto-hudi/pom.xml]: Since we just worked on another connector, and have a pending PR on a connector for Alluxio metadata service (#13986), you may find some of our experience useful. 

We initially started down the path of having our own connector, similar to how you have `presto-hudi`, it gives us a lot more flexibility, and allows us to use the additional statistics to affect query plan, but the cost of that is we had to write more code to get a working version. We switched to just implementing a meta-store in `presto-hive-metastore`, it was a much simpler path, but we lose some of that flexibility. In your case, you may find merging into `presto-hive` restricting as well, because hudi is more than just a file type.

Comment (bhasudha) [reply presto-hudi/pom.xml]: Thanks @haoyuan for pointing this. I ll take a look at this.

**Comment by bhasudha** [comment]
[UPDATE] After further discussions, we decided we will move this implementation inside presto-hive instead of a seperate connector for now. Presto-hive can support HoodieParquetInputFormat as a first class citizen eventually. Depending on future requirements, moving hudi to a seperate connector can be considered. I will update this PR and send out the changes soon. @arhimondr @wenleix @highker

**Comment by bhasudha** [comment]
@arhimondr  I updated the PR based on our last discussion. Please review when possible. I noticed that the Travis build failed in presto-cassandra module. Looks like it might be a transient connection issue. Would you be able to rekick the build ? I cant see an option for restarting the build in Travis. Thanks much!

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 11]
```diff
@@ -111,6 +116,7 @@
     private final ConcurrentLazyQueue<HivePartitionMetadata> partitions;
     private final Deque<Iterator<InternalHiveSplit>> fileIterators = new ConcurrentLinkedDeque<>();
     private final boolean schedulerUsesHostAddresses;
+    private final Supplier<HoodieROTablePathFilter> pathFilterSupplier;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: nit: let's call it `hoodiePathFilterSupplier`

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 12]
```diff
@@ -371,7 +376,7 @@ private void invokeNoMoreSplitsIfNecessary()
 
             return addSplitsToSource(splits, splitFactory);
         }
-
+        PathFilter pathFilter = (isHudiInputFormat(inputFormat) ? pathFilterSupplier.get() : path1 -> true);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: nit: parenthesis are extra

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java [hunk 13]
```diff
@@ -129,26 +136,39 @@ private FileStatusIterator(Path path, ListDirectoryOperation listDirectoryOperat
...
             catch (IOException e) {
                 throw processException(e);
             }
         }
 
+        @Override
+        public boolean hasNext()
+        {
+            computeNext();
+            return (next != null);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java]: nit: parenthesis are extra

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java [hunk 14]
```diff
@@ -92,14 +96,14 @@ protected HiveFileInfo computeNext()
...
+            remoteIterator = getLocatedFileStatusRemoteIterator(paths.removeFirst(), pathFilter);
         }
     }
 
-    private Iterator<HiveFileInfo> getLocatedFileStatusRemoteIterator(Path path)
+    private Iterator<HiveFileInfo> getLocatedFileStatusRemoteIterator(Path path, PathFilter pathFilter)
     {
         try (TimeStat.BlockTimer ignored = namenodeStats.getListLocatedStatus().time()) {
-            return new FileStatusIterator(path, listDirectoryOperation, namenodeStats);
+            return new FileStatusIterator(path, listDirectoryOperation, namenodeStats, pathFilter);
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java]: nit: Did you consider using the `com.google.common.collect.Iterators#filter` instead?

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+    private String pathFilterString = "path_filter_test_file_";
+    private Configuration hadoopConf;
+    private Random random;
+    private HiveFileIterator.ListDirectoryOperation listDirectoryOperation;
+    private static final String RANDOM_FILE_NAME_SALT_STRING = "abcdefghijklmnopqrstuvwxyz";
+
+    @BeforeClass
+    private void setup()
+    {
+        random = new Random();
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: `Random` is not thread safe, and the tests are ran in multiple threads. Please use `ThreadLocalRandom` instead (inline `ThreadLocalRandom.current()` everywhere)

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+import static com.google.common.io.Files.createTempDir;
+import static org.testng.Assert.assertEquals;
+
+public class TestHiveFileIterator
+{
+    private String pathFilterString = "path_filter_test_file_";
+    private Configuration hadoopConf;
+    private Random random;
+    private HiveFileIterator.ListDirectoryOperation listDirectoryOperation;
+    private static final String RANDOM_FILE_NAME_SALT_STRING = "abcdefghijklmnopqrstuvwxyz";
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: Please order your fields as CONSTANTS -> fields (move this field to the top, as split it with a space).

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.Random;
+
+import static com.facebook.presto.hive.util.HiveFileIterator.NestedDirectoryPolicy.IGNORED;
+import static com.facebook.presto.hive.util.HiveFileIterator.NestedDirectoryPolicy.RECURSE;
+import static com.google.common.io.Files.createTempDir;
+import static org.testng.Assert.assertEquals;
+
+public class TestHiveFileIterator
+{
+    private String pathFilterString = "path_filter_test_file_";
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: this can be constant (`private static final ...`)

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+    private Configuration hadoopConf;
+    private Random random;
+    private HiveFileIterator.ListDirectoryOperation listDirectoryOperation;
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: Static import `ListDirectoryOperation`

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 19]
```diff
@@ -0,0 +1,223 @@
...
+    private void setup()
+    {
+        random = new Random();
+        hadoopConf = new Configuration();
+        hadoopConf.set("fs.hdfs.impl", org.apache.hadoop.hdfs.DistributedFileSystem.class.getName());
+        hadoopConf.set("fs.file.impl", org.apache.hadoop.fs.LocalFileSystem.class.getName());
+        listDirectoryOperation = path ->
+        {
+            FileSystem fs = path.getFileSystem(hadoopConf);
+            return new HadoopDirectoryLister.HadoopFileInfoIterator(fs.listLocatedStatus(path));
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: static import `HadoopFileInfoIterator`

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+    private Configuration hadoopConf;
+    private Random random;
+    private HiveFileIterator.ListDirectoryOperation listDirectoryOperation;
+    private static final String RANDOM_FILE_NAME_SALT_STRING = "abcdefghijklmnopqrstuvwxyz";
+
+    @BeforeClass
+    private void setup()
+    {
+        random = new Random();
+        hadoopConf = new Configuration();
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: nit: Please nullify this field in `@AfterClass(alwaysRun = true)`

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+    }
+
+    private void deleteTestDir(File rootDir)
+    {
+        if (rootDir != null && rootDir.exists()) {
+            delete(rootDir);
+        }
+    }
+
+    private void delete(File file)
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: You can use `MoreFiles.deleteRecursively(tempDir.toPath(), ALLOW_INSECURE)` as in other test files (please import statically).

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 22]
```diff
@@ -0,0 +1,223 @@
...
+        int actualCount = getFileCount(hiveFileIterator);
+        assertEquals(actualCount, 9);
+
+        // cleanup
+        deleteTestDir(rootDir);
+    }
+
+    private void deleteTestDir(File rootDir)
+    {
+        if (rootDir != null && rootDir.exists()) {
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: It should never be null

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 23]
```diff
@@ -0,0 +1,223 @@
...
+                delete(child);
+            }
+        }
+        file.delete();
+    }
+
+    private void createFiles(String basePath, int numFiles, boolean matchPathFilter) throws IOException
+    {
+        for (int i = 0; i < numFiles; i++) {
+            new File(basePath).mkdirs();
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: Move it out of the loop

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+            String fileName;
+            if (matchPathFilter) {
+                fileName = pathFilterString + "_" + i;
+            }
+            else {
+                fileName = getRandomStringOfLen(10);
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: does it really have to be generated randomly? Can it be simply `"__not_matched_prefix_" + i`  instead?

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+            }
+            new File(basePath + File.separator + fileName).createNewFile();
+        }
+    }
+
+    private List<File> createDirs(String basePath, int numDirectories)
+    {
+        List<File> directories = new ArrayList<>();
+        for (int i = 0; i < numDirectories; i++) {
+            String dirName = basePath + File.separator + pathFilterString + "_" + getRandomStringOfLen(5);
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: Can `i` simply be used instead of `getRandomStringOfLen(5)`?

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+    private String getRandomStringOfLen(int len)
+    {
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < len; i++) {
+            sb.append(RANDOM_FILE_NAME_SALT_STRING.charAt(random.nextInt(RANDOM_FILE_NAME_SALT_STRING.length())));
+        }
+        return sb.toString();
+    }
+
+    public int getFileCount(HiveFileIterator hiveFileIterator)
```
Comment (arhimondr) [inline presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: `Iterators#size` can be used instead

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 11 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 12 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: fixed

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java (hunk 13 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 15 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: removed random usage

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 16 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 17 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: done

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 18 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 19 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 20 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: sure

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 21 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 22 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 24 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 25 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: sure

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 26 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: sure

File: presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java (hunk 23 above)
Comment (bhasudha) [reply presto-hive/src/test/java/com/facebook/presto/hive/util/TestHiveFileIterator.java]: fixed

File: presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java (hunk 14 above)
Comment (bhasudha) [reply presto-hive/src/main/java/com/facebook/presto/hive/util/HiveFileIterator.java]: Fixed
